archive/extract.go

```diff
@@ -15,10 +16,35 @@ type PathMode struct {
 	Mode os.FileMode
 }
 
+var (
+	umaskLock      sync.Mutex
+	extractCounter int
+	originalUmask  int
+)
+
+func setUmaskIfNeeded() {
+	umaskLock.Lock()
+	extractCounter++
+	if extractCounter == 1 {
+		originalUmask = setUmask(0)
+	}
+	umaskLock.Unlock()
+}
+
+func unsetUmaskIfNeeded() {
+	umaskLock.Lock()
+	extractCounter--
+	if extractCounter == 0 {
+		_ = setUmask(originalUmask)
+	}
+	umaskLock.Unlock()
```

Comment:
Same comment as the other. A `defer` feels appropriate here to me.

Reply:
Please see https://github.com/buildpacks/lifecycle/pull/730/commits/901a5d6146ffee65c3eb3a9c460df2446af4f6ec.